File: main/matrix-hr/matrix-hr.py
#!/usr/bin/env python3

"""
"Matrix" graph theory problem from HackerRank - union-find solution.
"""

import collections


class DisjointSets:
    """
    Disjoint-set-union data structure where each component is inactive
    (uninfected) or active (infected). The union of components is active when
    either was already active. The purpose of this data structure is to detect
    each time two already-active components are united.
    """

    __slots__ = ('_parents', '_ranks', '_active')

    def __init__(self, count):
        """Makes the specified number of singletons."""
        if count < 0:
            raise ValueError('negative count makes no sense')

        self._parents = list(range(count))
        self._ranks = [0] * count
        self._active = [False] * count

    def activate(self, vertex):
        """Marks the specified element as an active vertex."""
        self._check_exists(vertex)
        self._active[vertex] = True

    def union(self, vertex1, vertex2):
        """Unite by rank. Returns true iff separate active sets were joined."""
        self._check_exists(vertex1)
        self._check_exists(vertex2)

        # Find the ancestors and stop if they are already the same.
        vertex1 = self._find_set(vertex1)
        vertex2 = self._find_set(vertex2)
        if vertex1 == vertex2:
            return False

        # Unite by rank.
        if self._ranks[vertex1] < self._ranks[vertex2]:
            return self._join(vertex2, vertex1)
        if self._ranks[vertex1] == self._ranks[vertex2]:
            self._ranks[vertex1] += 1
        return self._join(vertex1, vertex2)

    def _find_set(self, vertex):
        """Finds the representative element, doing full path compression."""
        # Find the ancestor.
        leader = vertex
        while leader != self._parents[leader]:
            leader = self._parents[leader]

        # Compress the path.
        while vertex != leader:
            self._parents[vertex], vertex = leader, self._parents[vertex]

        return leader

    def _join(self, parent, child):
        """Joins the child to the parent. Returns true iff both were active."""
        self._parents[child] = parent

        if self._active[child]:
            if self._active[parent]:
                return True
            self._active[parent] = True

        return False

    def _check_exists(self, vertex):
        """Raises an exception if the vertex does not exist."""
        if not 0 <= vertex < len(self._parents):
            raise IndexError('vertex does not exist (out of range)')


Edge = collections.namedtuple('Edge', ('vertex1', 'vertex2', 'weight'))
Edge.__doc__ = """An edge in a weighted undirected graph."""


def read_value():
    """Reads a line of input as a single integer."""
    return int(input())


def read_record():
    """Reads a line of input as a sequence of integers."""
    return map(int, input().split())


def read_edges_sorted(edge_count):
    """Reads and reverse-weight sorts the specified number of edges."""
    edges = [Edge._make(read_record()) for _ in range(edge_count)]
    edges.sort(key=lambda edge: edge.weight, reverse=True)
    return edges


def read_vertices(count, active_count):
    """Reads one active vertex per line. Makes singletons for all vertices."""
    sets = DisjointSets(count)
    for _ in range(active_count):
        sets.activate(read_value())
    return sets


def run():
    """Solves the problem as described on standard input."""
    city_count, machine_count = read_record()
    if city_count <= 0:
        raise ValueError('must have at least one city')
    if machine_count < 0:
        raise ValueError("can't have negatively many machines")
    if city_count < machine_count:
        raise ValueError("can't have more machines than cities")

    edges = read_edges_sorted(city_count - 1)
    sets = read_vertices(city_count, machine_count)
    total_bridge_weight = 0
    bridge_count = 0  # for error checking and debugging

    for vertex1, vertex2, weight in edges:
        if sets.union(vertex1, vertex2):
            total_bridge_weight += weight
            bridge_count += 1

    assert bridge_count == machine_count - 1
    print(total_bridge_weight)


if __name__ == '__main__':
    run()

File: main/matrix-hr/matrix-hr.cpp
#ifdef NDEBUG
#error Assertions unexpectedly disabled.
#endif

#include <algorithm>
#include <cassert>
#include <cstddef>
#include <functional>
#include <iostream>
#include <iterator>
#include <numeric>
#include <vector>

namespace {
    // Disjoint-set-union data structure where each component is inactive
    // (uninfected) or active (infected). The union of components is active
    // when either was already active. The purpose of this data structure is to
    // detect each time two already-active components are united.
    class DisjointSets {
    public:
        explicit DisjointSets(int count);

        // Marks the specified element as an active vertex.
        void activate(int vertex) noexcept;

        // Union by rank. Returns true iff separate active sets were joined.
        bool unite(int vertex1, int vertex2) noexcept;
    
    private:
        enum class State : bool { inactive, active };

        static std::size_t convert_count(int count) noexcept;

        explicit DisjointSets(std::size_t count);

        int count() const noexcept;

        bool exists(int vertex) const noexcept;

        // Finds the representative element, performing full path compression.
        int find_set(int vertex) noexcept;

        // Joins the child to the parent. Returns true iff both were active.
        bool join(int parent, int child) noexcept;

        std::vector<int> parents_;
        std::vector<int> ranks_;
        std::vector<State> states_;
    };

    inline std::size_t DisjointSets::convert_count(int count) noexcept
    {
        assert(0 <= count);
        return static_cast<std::size_t>(count);
    }

    DisjointSets::DisjointSets(const std::size_t count)
        : parents_(count), ranks_(count), states_(count)
    {
        std::iota(begin(parents_), end(parents_), 0);
    }

    DisjointSets::DisjointSets(const int count)
        : DisjointSets{convert_count(count)}
    {
    }

    inline int DisjointSets::count() const noexcept
    {
        return static_cast<int>(parents_.size());
    }

    inline bool DisjointSets::exists(const int vertex) const noexcept
    {
        return 0 <= vertex && vertex < count();
    }

    inline void DisjointSets::activate(const int vertex) noexcept
    {
        assert(exists(vertex));
        states_[vertex] = State::active;
    }

    bool DisjointSets::unite(int vertex1, int vertex2) noexcept
    {
        assert(exists(vertex1));
        assert(exists(vertex2));

        // Find the ancestors and stop of they are already the same.
        vertex1 = find_set(vertex1);
        vertex2 = find_set(vertex2);
        if (vertex1 == vertex2) return false;

        // Union by rank.
        if (ranks_[vertex1] < ranks_[vertex2]) return join(vertex2, vertex1);
        if (ranks_[vertex1] == ranks_[vertex2]) ++ranks_[vertex1];
        return join(vertex1, vertex2);
    }

    int DisjointSets::find_set(int vertex) noexcept
    {
        // Find the ancestor.
        auto leader = vertex;
        while (leader != parents_[leader]) leader = parents_[leader];

        // Compress the path.
        while (vertex != leader) {
            const auto parent = parents_[vertex];
            parents_[vertex] = leader;
            vertex = parent;
        }

        return leader;
    }

    bool DisjointSets::join(const int parent, const int child) noexcept
    {
        parents_[child] = parent;

        if (states_[child] == State::active) {
            if (states_[parent] == State::active) return true;
            states_[parent] = State::active;
        }

        return false;
    }

    // An edge in an weighted undirected graph. Edges compare by their weights.
    struct Edge {
        int vertex1;
        int vertex2;
        long long weight;
    };

    constexpr bool operator<(const Edge& lhs, const Edge& rhs) noexcept
    {
        return lhs.weight < rhs.weight;
    }

    constexpr bool operator>(const Edge& lhs, const Edge& rhs) noexcept
    {
        return lhs.weight > rhs.weight;
    }

    inline std::istream& operator>>(std::istream& is, Edge& edge)
    {
        return is >> edge.vertex1 >> edge.vertex2 >> edge.weight;
    }

    std::vector<Edge> read_edges_sorted(const int edge_count)
    {
        std::vector<Edge> edges (static_cast<std::size_t>(edge_count));

        std::copy_n(std::istream_iterator<Edge>{std::cin},
                    edge_count, begin(edges));
        
        std::sort(begin(edges), end(edges), std::greater<>{});
        return edges;
    }

    DisjointSets read_vertices(const int count, int active_count)
    {
        DisjointSets sets {count};

        while (active_count-- > 0) {
            int vertex {};
            std::cin >> vertex;
            assert(std::cin);
            sets.activate(vertex);
        }

        return sets;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    int city_count {}, machine_count {};
    std::cin >> city_count >> machine_count;
    
    assert(0 < city_count);
    assert(0 <= machine_count);
    assert(machine_count <= city_count);

    const auto edges = read_edges_sorted(city_count - 1);
    auto sets = read_vertices(city_count, machine_count);
    auto total_bridge_weight = 0LL;
    auto bridge_count = 0; // for error checking and debugging

    for (const auto& edge : edges) {
        if (sets.unite(edge.vertex1, edge.vertex2)) {
            total_bridge_weight += edge.weight;
            ++bridge_count;
        }
    }
    
    assert(bridge_count == machine_count - 1);
    std::cout << total_bridge_weight << '\n';
}
